internal/services/tesla_api_service.go

```diff
@@ -0,0 +1,130 @@
+package services
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/DIMO-Network/devices-api/internal/config"
+	"github.com/rs/zerolog"
+	"golang.org/x/oauth2"
+)
+
+//go:generate mockgen -source tesla_api_service.go -destination mocks/tesla_api_service_mock.go
+type TeslaAPIService interface {
+	CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error)
+	GetVehicles(token, region string) ([]TeslaVehicle, error)
+}
+
+var teslaScopes = []string{"openid offline_access", "user_data", "vehicle_device_data", "vehicle_cmds", "vehicle_charging_cmds", "energy_device_data", "energy_device_data", "energy_cmds"}
+
+type GetVehiclesResponse struct {
+	Response []TeslaVehicle `json:"response"`
+}
+
+type TeslaError struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+	ReferenceID      string `json:"ReferenceID"`
+}
+
+type TeslaAuthCodeResponse struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	IDToken      string `json:"id_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	State        string `json:"state"`
+	TokenType    string `json:"token_type"`
+}
+
+type teslaAPIService struct {
+	Settings   *config.Settings
+	HTTPClient *http.Client
+	log        *zerolog.Logger
+}
+
+func NewTeslaAPIService(settings *config.Settings, logger *zerolog.Logger) TeslaAPIService {
+	return &teslaAPIService{
+		Settings:   settings,
+		HTTPClient: &http.Client{},
+		log:        logger,
+	}
+}
+
+func (t *teslaAPIService) CompleteTeslaAuthCodeExchange(authCode, redirectURI, region string) (*TeslaAuthCodeResponse, error) {
+	conf := oauth2.Config{
+		ClientID:     t.Settings.Tesla.ClientID,
+		ClientSecret: t.Settings.Tesla.ClientSecret,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: t.Settings.Tesla.TokenURL,
+		},
+		RedirectURL: redirectURI,
+		Scopes:      teslaScopes,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+
+	tok, err := conf.Exchange(ctx, authCode, oauth2.SetAuthURLParam("audience", fmt.Sprintf("https://fleet-api.prd.%s.vn.cloud.tesla.com", region)), oauth2.SetAuthURLParam("grant_type", "authorization_code"))
+	if err != nil {
+		var e *oauth2.RetrieveError
+		errors.As(err, &e)
+		return nil, fmt.Errorf("error occurred completing authorization: %s", e.ErrorDescription)
+	}
+
+	return &TeslaAuthCodeResponse{
+		AccessToken:  tok.AccessToken,
+		RefreshToken: tok.RefreshToken,
+		ExpiresIn:    int(tok.Expiry.Unix()),
```

Comment:
I don't think this is right. `tok.Expiry` is the time at which this thing expires. It's not `expires_in`. It's `time.Now() + expires_in`.

Reply:
<img width="2372" alt="image" src="https://github.com/DIMO-Network/devices-api/assets/111298305/76f00b4c-43ab-47f1-8f69-4cf4aba690a9">

Not sure I follow your meaning, but this is what the API response from Tesla looks like

Reply:
I'm not sure I understand what you mean. Also upon checking again, seems tesla returns expires_in and not expiry, but the oauth library seems to have expiry instead.

Reply:
I think the `oauth2` library computes `Expiry` as `time.Now() + expires_in`, where `expires_in` is coming from the token.